app/lib/action_view/component.rb

```diff
@@ -26,45 +28,55 @@ class Component < ActionView::Base
     # <%= render MyComponent.new(title: "greeting") do %>world<% end %>
     # returns:
     # <span title="greeting">Hello, world!</span>
-    #
-    def render_in(view_context, *args, &block)
-      self.class.compile
-      @content = view_context.capture(&block) if block_given?
-      validate!
-      call
-    end
 
-    def initialize(*); end
+    # @see https://github.com/devnacho/mountain_view/blob/master/lib/mountain_view/presenter.rb#L65
+    module ViewContext
+      attr_reader :_component
 
-    def self.compile
-      @compiled ||= nil
-      return if @compiled
+      def inject_component_context(component)
+        @_component = component
 
-      class_eval(
-        "def call; @output_buffer = ActionView::OutputBuffer.new; " +
-        ActionView::Template::Handlers::ERB.erb_implementation.new(template, trim: true).src +
-        "; end"
-      )
+        protected_methods = ActionView::Component.public_methods(false)
+        methods = component.public_methods(false) - protected_methods
 
-      @compiled = true
-    end
+        methods.each do |meth|
+          next if self.class.method_defined?(meth)
+          next unless _component.respond_to?(meth)
 
-    def self.template
-      filename = self.instance_method(:initialize).source_location[0]
+          self.class.delegate meth, to: :_component
+        end
+      end
+    end
 
-      raise NotImplementedError.new("Subclasses of ActionView::Component must implement #initialize") if filename == __FILE__
+    def render_in(context, *_args)
+      context.extend ViewContext
+      context.inject_component_context self
 
-      filename_without_extension = filename[0..-(File.extname(filename).length + 1)]
+      @content = yield if block_given?
```

Comment:
Help needed with passing the context down to the block.

See the `Issues::Badge` component, which has a `state` method.
See the `app/components/issues/badge.html.erb` which calls the `state` method in the block.

Reply:
@tomasc right now we're using the `capture` helper, but I'm not sure how we might do something similar here.